allocation_manager/service/unit.go

```diff
@@ -46,18 +49,56 @@ metricLicenseReleaseReason = promauto.NewCounterVec(prometheus.CounterOpts{
 	},
 )
 */
+
+unitCounter = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "unit_operations_total",
+		Help: "Total number of operations performed on units",
+	},
+	[]string{"kind", "unit", "operation"},
+)
 )
 
+func init() {
+	fmt.Println("In init")
+	prometheus.MustRegister(unitCounter)
+}
+
+func (u unit) DoOperation(operationName string) {
+	// name := u.name
+	name := u.Topology.Name
+	unitCounter.With(prometheus.Labels{
+		"kind": fmt.Sprintf("%T", u),
+		"unit": name,
+		"operation": operationName}).Inc()
+	fmt.Printf("Operation '%s' performed on unit: %s %T\n", operationName, name, u)
+	// fmt.Println("Counter:", unitCounter.Counter)
+}
+
 type unit struct { // store topologies describing actual hardware
 	Health     apb.Health   // Health status of hardware
 	Topology   apb.Topology // hardware actual configuration
 	Invocation *invocation  // request for a Unit allocation
 }
 
+func NewUnit(topo apb.Topology) *unit {
+	u := new(unit)
+	u.Topology = topo
+	fmt.Println("New unit created", topo)
+	return u
+}
+
+// Returns whether this Unit was successfully released.
+func (u *unit) Release(inv *invocation) bool {
```

Comment:
I guess it wasn't obvious; the Release rpc calls Forget() below.

Reply:
Also, ExpireAllocations() is another way that Units are released, but it's not caused by client action (rather by inaction).